astropy/units/physical.py

```diff
@@ -1,133 +1,473 @@
+    True
+
+    Multiplication, division, and exponentiation are enabled so that
+    physical types may be used for dimensional analysis.
+
+    >>> length = u.pc.physical_type
+    >>> area = (u.cm ** 2).physical_type
+    >>> length * area
+    PhysicalType('volume')
+    >>> area / length
+    PhysicalType('length')
+    >>> length ** 3
+    PhysicalType('volume')
+
+    Unknown physical types are labelled as ``"unknown"``.
+
+    >>> (u.s ** 13).physical_type
+    PhysicalType('unknown')
+
+    Dimensional analysis may be performed for unknown physical types too.
+
+    >>> length_to_19th_power = (u.m ** 19).physical_type
+    >>> length_to_20th_power = (u.m ** 20).physical_type
+    >>> length_to_20th_power / length_to_19th_power
+    PhysicalType('length')
+    """
+
+    def __init__(self, unit, physical_types):
+        self._unit = _replace_temperatures_with_kelvin(unit)
+        self._physical_type_id = self._unit._get_physical_type_id()
+        self._physical_type = _standardize_physical_type_names(physical_types)
+        self._physical_type_list = sorted(self._physical_type)
+
+    def __iter__(self):
+        yield from self._physical_type_list
+
+    def __eq__(self, other):
+        """
+        Return `True` if ``other`` represents a physical type that is
+        consistent with the physical type of the `PhysicalType` instance.
+        """
+        if isinstance(other, PhysicalType):
+            return self._physical_type_id == other._physical_type_id
+        elif isinstance(other, str):
+            other = _standardize_physical_type_names(other)
+            return other.issubset(self._physical_type)
+        else:
+            return NotImplemented
+
+    def __ne__(self, other):
+        equality = self.__eq__(other)
+        return not equality if isinstance(equality, bool) else NotImplemented
+
+    @property
+    def _name_string_as_ordered_set(self):
+        return "{" + str(self._physical_type_list)[1:-1] + "}"
+
+    def __repr__(self):
+        if len(self._physical_type) == 1:
+            names = "'" + self._physical_type_list[0] + "'"
+        else:
+            names = self._name_string_as_ordered_set
+        return f"PhysicalType({names})"
+
+    def __str__(self):
+        if len(self._physical_type) == 1:
+            return self._physical_type_list[0]
+        else:
+            return self._name_string_as_ordered_set
+
+    @staticmethod
+    def _dimensionally_compatible_unit(obj):
+        """
+        If a unit is passed in, return that unit.  If a physical type is
+        passed in, return a unit that corresponds to that physical type.
+        If a real number is passed in, return a dimensionless unit.
+        Otherwise, return `NotImplemented`.
+        """
+        if isinstance(obj, core.UnitBase):
+            return _replace_temperatures_with_kelvin(obj)
+        elif isinstance(obj, PhysicalType):
+            return obj._unit
+        elif isinstance(obj, numbers.Real):
```

Comment:
I'm a bit unsure why we allow numbers here; for dimensional analysis, that should not be useful.

Reply:
That's a good question!  

My first use case for allowing numbers is that there might be times when someone would want to do an operation like:
```Python
>>> length = u.m.physical_type
>>> 1 / length
PhysicalType('wavenumber')
```
which would be equivalent to `length ** -1`.

My second use case is that if I were introducing the idea of dimensional analysis to someone, I might use an example like
```Python
>>> length * 42
PhysicalType('length')
```
to show that when doing dimensional analysis, multiplying by a constant will not change the physical type.  This would also be equivalent to `length ** 1`.  

I was also thinking this would be reasonable since SI defines "one" as a unit that is implicitly present.  (This make me wonder if an alternative acceptable name for `u.dimensionless_unscaled` would be `u.one`, but it's a few years too late to suggest that.  Nevermind, I just looked and `u.one` is defined as `u.dimensionless_unscaled`.  😹 )

In any case, are these use cases sufficient motivation to allow these operations?

Reply:
Seems like good enough motivation to me!

Reply:
I must admit I'm not convinced, especially not with the multiplication with a number: really, what is meant by `42 * PhysicalType('length')`? To me, the physical type is very strictly tied to units, and for units, `1/unit` is treated differently from `unit**-1` (quantities do not have physical type).

That said, the `1/phystype` example is something I could see myself doing...  So,  while my tendency would still be to start strict and not allow it, I am also OK with for now allowing just that one (and only integer 1, i.e., `if type(obj) is int and obj == 1`)

Reply:
p.s. In that case, `1` should also be allowed as input for `PhysicalType`. Indeed, I could live with allowing string inputs here, and thus just using the initializer to parse `obj`.  This would be consistent with what we do with units, where `u.m / 's'` is allowed.

Reply:
> That said, the `1/phystype` example is something I could see myself doing... So, while my tendency would still be to start strict and not allow it, I am also OK with for now allowing just that one (and only integer 1, i.e., `if type(obj) is int and obj == 1`)

Sure, that sounds like a fair compromise.  Definitely something I can live with!  On the chance that there are more use cases that come up in the future, this could be revisited.

> p.s. In that case, `1` should also be allowed as input for `PhysicalType`. Indeed, I could live with allowing string inputs here, and thus just using the initializer to parse `obj`. This would be consistent with what we do with units, where `u.m / 's'` is allowed.

Good ideas!